python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py

```diff
@@ -528,28 +529,116 @@ def is_graph_backed_asset(self) -> bool:
+                asset_entry.last_materialization_storage_id is None
+                and asset_entry.last_failed_to_materialize_storage_id is None
+            ):
+                # never materialized
+                return GrapheneAssetHealthStatus.UNKNOWN
+            if asset_entry.last_failed_to_materialize_storage_id is None:
+                # last_materialization_record must be non-null, therefore the asset successfully materialized
+                return GrapheneAssetHealthStatus.HEALTHY
+            elif asset_entry.last_materialization_storage_id is None:
+                # last_failed_to_materialize_record must be non-null, therefore the asset failed to materialize
+                return GrapheneAssetHealthStatus.DEGRADED
+
+            if (
+                asset_entry.last_materialization_storage_id
+                > asset_entry.last_failed_to_materialize_storage_id
+            ):
+                # latest materialization succeeded
+                return GrapheneAssetHealthStatus.HEALTHY
+            # latest materialization failed
+            return GrapheneAssetHealthStatus.DEGRADED
+
         else:
-            # commpute the status based on the asset key table (or using asset status cache for partitioned)
-            pass
+            # check what we can before fetching the run
+            fallback_status = (
+                GrapheneAssetHealthStatus.UNKNOWN
+                if asset_entry.last_materialization is None
+                else GrapheneAssetHealthStatus.HEALTHY
+            )
+            if asset_entry.last_run_id is None:
+                if asset_entry.last_materialization is None:
+                    # never materialized
+                    return GrapheneAssetHealthStatus.UNKNOWN
+                else:
+                    # last materialization was manually reported
+                    return GrapheneAssetHealthStatus.HEALTHY
 
-        return GrapheneAssetHealthStatus.UNKNOWN
+            assert asset_entry.last_run_id is not None
+            if (
+                asset_entry.last_materialization is not None
+                and asset_entry.last_run_id == asset_entry.last_materialization.run_id
+            ):
+                # latest materialization succeeded in the latest run
+                return GrapheneAssetHealthStatus.HEALTHY
+            run_record = await RunRecord.gen(graphene_info.context, asset_entry.last_run_id)
+            if run_record is None or not run_record.dagster_run.is_finished:
+                return fallback_status
+            run_end_time = check.not_none(run_record.end_time)
+            if (
+                asset_entry.last_materialization
+                and asset_entry.last_materialization.timestamp > run_end_time
+            ):
+                # latest materialization was reported manually
+                return GrapheneAssetHealthStatus.HEALTHY
+            if run_record.dagster_run.is_failure:
+                return GrapheneAssetHealthStatus.DEGRADED
```

Comment:
The similar logic for displaying status like this for unpartitioned assets used to live on the frontend right? This seems like a better place for it

Reply:
yeah this branch of code when we don't have failure events is based on the FE implementation for showing status